### PR TITLE
Dex happy path

### LINKED ISSUE #403: Dex: Happy path flow

### LINKED ISSUE #445: DEX Example - redeemLiquidity method should run correctly
# Description

The current [redeemLiquidity](https://github.com/o1-labs/snarkyjs/blob/200be0ca0f575e08af2a62e1a3a001e7f53fd787/src/examples/zkapps/dex/dex.ts#L107) method on the DEX example throws an error from OCaml. 

Further, inspection is needed to understand why the error is being thrown and how to fix it.

### LINKED ISSUE #482: StackOverflow error caused by cloning `lazyAuthorization` inside `toAuxiliary`
# Description
When calling `this.experimental.authorize(accountUpdate)` inside a smart contract method, a stack overflow error occurs. This happens because inside `toAuxiliary`, the function will call `cloneCircuitValue` on `a.lazyAuthorization` and recursively call `cloneCircuitValue` until it blows up the stack.

See this line here: https://github.com/o1-labs/snarkyjs/blob/ecffd00ecb39f6aa24f2ead47e51eda570ab8071/src/lib/account_update.ts#L1095

This was confirmed by commenting out the `cloneCircuitValue` call and ensuring that no stack overflow error occurs.

### PR DESCRIPTION
**Pull request by MartinMinkov**
This PR finishes the happy path of the first DEX version, without proofs. Proofs currently run into an exotic Pickles error when executing "redeemLiquidity". We checked that the account update structures created by the DEX methods make sense and are as expected.

This also
* makes some changes to how Circuit.witness works,
* adds a `Circuit.log` function (use like console.log) to remove `Circuit.asProver` boilerplate when debugging,
* makes adding account updates to the tx more explicit in some cases where it happened in a very hidden way, causing bugs
* Closes #482
* Closes #403 
* Closes #445

### CODE REVIEW

File: src/examples/zkapps/dex/arbitrary_token_interaction.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+console.log('compile (token)...');
+await TokenContract.compile();
+
+let tokenX = new TokenContract(addresses.tokenX);
+
+console.log('deploy & init token contracts...');
+tx = await Mina.transaction(userKey, () => {
+  // pay fees for creating 2 token contract accounts, and fund them so each can create 1 account themselves
+  let feePayerUpdate = AccountUpdate.createSigned(userKey);
+  feePayerUpdate.balance.subInPlace(accountFee.mul(1));
```
Comment (Trivo25) [inline src/examples/zkapps/dex/arbitrary_token_interaction.ts:38]: Is there a reason for multiplying it times one? Probably not, just curious tho

File: src/examples/zkapps/dex/dex.ts [hunk 2]
```diff
@@ -58,6 +64,7 @@ class Dex extends SmartContract {
     let isXZero = dexXBalance.equals(UInt64.zero);
     let xSafe = Circuit.if(isXZero, UInt64.one, dexXBalance);
 
+    // FIXME
```
Comment (Trivo25) [inline src/examples/zkapps/dex/dex.ts:67]: is this important?

File: src/lib/circuit_value.ts [hunk 3]
```diff
@@ -1003,18 +1009,30 @@ Circuit.constraintSystem = function <T>(f: () => T) {
   return result;
 };
 
+Circuit.log = function (...args: any) {
```
[the remark this answers is not in the document]
Comment (Trivo25) [reply src/lib/circuit_value.ts:1012]: great addition

**Review by Trivo25 — APPROVED** [review]
lgtm, just left a few questions :)

File: src/examples/zkapps/dex/arbitrary_token_interaction.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+});
+await tx.prove();
+tx.sign([keys.tokenX]);
+await tx.send();
+
+console.log('arbitrary token minting...');
+tx = await Mina.transaction(userKey, () => {
+  // pay fees for creating user's token X account
+  AccountUpdate.createSigned(userKey).balance.subInPlace(accountFee.mul(1));
+  // 😈😈😈 mint any number of tokens to our account 😈😈😈
```
Comment (ymekuria) [inline src/examples/zkapps/dex/arbitrary_token_interaction.ts:49]: I love the comment emojis ha!

File: src/examples/zkapps/dex/run.ts [hunk 5]
```diff
@@ -9,9 +16,9 @@ import {
 } from './dex.js';
```
Comment (ymekuria) [inline src/examples/zkapps/dex/run.ts:17]: Was there a reason you created a separate `arbitrary_token_interaction.ts` file and this `run.ts` file? Is the goal to only execute the `run.ts` file in CI and not the `arbitrary_token_interaction.ts` file?

File: src/lib/circuit_value.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ymekuria) [reply src/lib/circuit_value.ts:1012]: This is awesome!

**Review by ymekuria — APPROVED** [review]
LGTM! I just had some minor comments/questions.

File: src/examples/zkapps/dex/arbitrary_token_interaction.ts (hunk 1 above)
Comment (mitschabaude) [reply src/examples/zkapps/dex/arbitrary_token_interaction.ts:38]: no there isn't :D it was copied from another place where it was multiplied by 2 I guess

File: src/examples/zkapps/dex/dex.ts (hunk 2 above)
Comment (mitschabaude) [reply src/examples/zkapps/dex/dex.ts:67]: yes! reminds us to fix it

File: src/examples/zkapps/dex/run.ts (hunk 5 above)
Comment (ymekuria) [reply src/examples/zkapps/dex/run.ts:17]: Ok that makes sense. Maybe we should document the file with this info so we don't forget later on.

**Pull request merged**